### ISSUE TITLE
Embedded Hazelcast 5 fail to startup if multiple instances are started in parallel with Istio envoy proxy enabled

### ISSUE THREAD

**Issue by michael-wirth** [hazelcast/hazelcast#24630]
**Describe the bug**
This bug is related to https://github.com/hazelcast/hazelcast/issues/22256
An application with an embedded Hazelcast fails to startup if multiple instances are started in parallel.

If added the `appProtocol: tcp` to the service spec but without any effects.

The joiner tried to connect to each possible member but fails to do a handshake with instances starting in parallel.
The Istio Proxy accepts the handshake connection but fails to forward it to the container (which is starting up).
Instead of dropping that member, Hazelcast is going in an infinite loop and tries to connect again and again. The error count in the logs below is already at 155 and is increasing as logger the app runs. The counter goes up to 2000 before the app ist restarted by Openshift.

How can Hazelcast be forced to drop a member after n failed requests?
What exactly was the change/solution in the referenced [issue](https://github.com/hazelcast/hazelcast/issues/22256)? The only "fix" I found was implemented in 2021, why is the issue assigned to release 5.3.0?

**Expected behavior**
Hazelcast should drop members it can't connect instead going into an infinite loop.

**Additional context**

Versions:
Java: 17.0.7+7
Spring Boot 3.1.0
Hazelcast 5.3.0
Openshift 4.10
Kubernetes 1.23.12+8a6bfe4
maistra-version: 2.1.6
Istio 1.14.5
Envoy Proxy 1.22.9

```log
2023-05-24T09:21:45.791+02:00 TRACE [ebcontainer,,,] 1 --- [hz.default.cached.thread-8] c.h.i.server.tcp.TcpServerConnector      : [10.105.4.212]:5701 [imp2] [5.3.0] Starting to connect to [10.105.3.103]:5701
2023-05-24T09:21:45.791+02:00 TRACE [ebcontainer,,,] 1 --- [hz.default.IO.BalancerThread] c.h.i.n.nio.iobalancer.IOBalancer        : [10.105.4.212]:5701 [imp2] [5.3.0] Adding pipelines: NioChannel{/[0:0:0:0:0:0:0:0]:58605->null}.inboundPipeline, NioChannel{/[0:0:0:0:0:0:0:0]:58605->null}.outboundPipeline
2023-05-24T09:21:45.791+02:00  INFO [ebcontainer,,,] 1 --- [hz.default.cached.thread-8] c.h.i.server.tcp.TcpServerConnector      : [10.105.4.212]:5701 [imp2] [5.3.0] Connecting to /10.105.3.103:5701, timeout: 10000, bind-any: true
2023-05-24T09:21:45.791+02:00 TRACE [ebcontainer,,,] 1 --- [hz.default.cached.thread-8] c.h.internal.networking.nio.NioChannel   : Successfully connected to: /10.105.3.103:5701 using socket Socket[addr=/10.105.3.103,port=5701,localport=58605]
2023-05-24T09:21:45.791+02:00 DEBUG [ebcontainer,,,] 1 --- [hz.default.cached.thread-8] c.h.i.s.tcp.TcpServerConnectionManager   : [10.105.4.212]:5701 [imp2] [5.3.0] Established socket connection between /[0:0:0:0:0:0:0:0]:58605 and /10.105.3.103:5701
2023-05-24T09:21:45.791+02:00 TRACE [ebcontainer,,,] 1 --- [hz.default.cached.thread-8] c.h.i.server.tcp.TcpServerConnector      : [10.105.4.212]:5701 [imp2] [5.3.0] Sending memberHandshake packet to [10.105.3.103]:5701
2023-05-24T09:21:45.791+02:00 DEBUG [ebcontainer,,,] 1 --- [hz.default.cached.thread-21] c.h.i.s.tcp.TcpServerConnectionManager   : [10.105.4.212]:5701 [imp2] [5.3.0] Connection to: [10.105.3.103]:5701 streamId:0 is already in progress
2023-05-24T09:21:45.792+02:00 DEBUG [ebcontainer,,,] 1 --- [hz.default.cached.thread-17] c.h.i.s.tcp.TcpServerConnectionManager   : [10.105.4.212]:5701 [imp2] [5.3.0] Connection to: [10.105.3.103]:5701 streamId:0 is already in progress
2023-05-24T09:21:45.792+02:00 DEBUG [ebcontainer,,,] 1 --- [hz.default.cached.thread-7] c.h.i.s.tcp.TcpServerConnectionManager   : [10.105.4.212]:5701 [imp2] [5.3.0] Connection to: [10.105.3.103]:5701 streamId:0 is already in progress
2023-05-24T09:21:45.793+02:00 DEBUG [ebcontainer,,,] 1 --- [hz.default.cached.thread-21] c.h.i.s.tcp.TcpServerConnectionManager   : [10.105.4.212]:5701 [imp2] [5.3.0] Connection to: [10.105.3.103]:5701 streamId:0 is already in progress
2023-05-24T09:21:45.793+02:00  INFO [ebcontainer,,,] 1 --- [hz.default.IO.thread-in-0] c.h.i.server.tcp.TcpServerConnection     : [10.105.4.212]:5701 [imp2] [5.3.0] Connection[id=9087, /[0:0:0:0:0:0:0:0]:58605->/10.105.3.103:5701, qualifier=null, endpoint=[10.105.3.103]:5701, remoteUuid=null, alive=false, connectionType=NONE, planeIndex=-1] closed. Reason: Connection closed by the other side
2023-05-24T09:21:45.793+02:00 TRACE [ebcontainer,,,] 1 --- [hz.default.IO.thread-in-0] .h.i.s.t.TcpServerConnectionErrorHandler : [10.105.4.212]:5701 [imp2] [5.3.0] An error occurred on connection to [10.105.3.103]:5701 Cause => java.io.EOFException {Remote socket closed!}, Error-Count: 154
2023-05-24T09:21:45.793+02:00  WARN [ebcontainer,,,] 1 --- [hz.default.IO.thread-in-0] .h.i.s.t.TcpServerConnectionErrorHandler : [10.105.4.212]:5701 [imp2] [5.3.0] Removing connection to endpoint [10.105.3.103]:5701 Cause => java.io.EOFException {Remote socket closed!}, Error-Count: 155   <<--- Why is this member not dropped with it failed for 155 times?
2023-05-24T09:21:45.795+02:00 DEBUG [ebcontainer,,,] 1 --- [hz.default.cached.thread-21] c.h.i.s.tcp.TcpServerConnectionManager   : [10.105.4.212]:5701 [imp2] [5.3.0] Connection to: [10.105.3.103]:5701 streamId:0 is not yet in progress
```

```yaml
kind: Service
apiVersion: v1
metadata:
  name: app-headless
  labels:
    app: app
    app.kubernetes.io/component: backend
    app.kubernetes.io/env: imp2
    app.kubernetes.io/managed-by: gitlab-kustomize
    app.kubernetes.io/name: app
    app.kubernetes.io/part-of: 1eb-imp2
    environment: imp2
    primary-port-name: tcp-management
  annotations:
    app.gitlab.com/app: eb-container
    app.gitlab.com/env: imp2
    app.kubernetes.io/build-version: build-2023.05.23T23.13
    app.kubernetes.io/version: 2321.0.0-SNAPSHOT
spec:
  clusterIP: None
  ipFamilies:
  - IPv4
  ports:
  - name: tcp-management
    protocol: TCP
    port: 9080
    targetPort: management
  - name: tcp-hazelcast
    protocol: TCP
    appProtocol: TCP
    port: 5701
    targetPort: hazelcast
  internalTrafficPolicy: Cluster
  clusterIPs:
  - None
  type: ClusterIP
  ipFamilyPolicy: SingleStack
  sessionAffinity: None
  selector:
    app: app
    app.kubernetes.io/component: backend
    app.kubernetes.io/env: imp2
    app.kubernetes.io/managed-by: gitlab-kustomize
    app.kubernetes.io/name: app
    app.kubernetes.io/part-of: 1eb-imp2
    environment: imp2
status:
  loadBalancer: {}
```

**Comment by hasancelik** [comment]
Hi @michael-wirth,

The configuration described in another issue could be useful for the problem you reported:
https://github.com/hazelcast/hazelcast/issues/26364#issuecomment-2189267973

Could you check that?
